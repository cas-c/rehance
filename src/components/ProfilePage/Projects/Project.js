import React from 'react';
import PropTypes from 'prop-types';

const Project = ({ data }) => (
    <a href={data.url} target='_blank' title={data.name} >
        <div className='card'>
            <header className='card-header'>
                <p className='card-header-title'>
                    { data.name }
                </p>
            </header>
            <div className='card-image project-cover'>
                <figure className='image is-4by3'>
                    {
                        data.covers && <img className='project-cover-img' src={data.covers.original} alt={data.name} />
                    }
                </figure>
            </div>
        </div>
    </a>
);

Project.propTypes = {
    data: PropTypes.shape({
        url: PropTypes.string,
        name: PropTypes.string,
        covers: PropTypes.shape({
            original: PropTypes.string
        })
    })
}

Project.defaultProps = {
    data: {}
}

export default Project;
