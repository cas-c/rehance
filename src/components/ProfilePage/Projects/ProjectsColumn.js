import React from 'react';
import PropTypes from 'prop-types';
import Project from './Project';

//todo: Can probably make 'Column' a more generic component (see MiniProfilesColumn)

const ProjectsColumn = ({ projects }) => (
    <div>
        <div className='columns'>
            {
                projects.length > 0 ?
                    projects.slice(0, 5).map((p, i) =>
                        <div className='column' key={`project-${i}-wrapper`}>
                            <Project data={p} key={`project-${i}`}/>
                        </div>
                    ) :
                    <div className='has-text-centered'> This user does not have any projects. </div>
            }
        </div>
        { projects.length > 5 && <div className='has-text-right'><a>View all...</a></div> }
    </div>
);

ProjectsColumn.propTypes = {
    projects: PropTypes.arrayOf(PropTypes.object)
};

ProjectsColumn.defaultProps = {
    projects: []
};

export default ProjectsColumn;
