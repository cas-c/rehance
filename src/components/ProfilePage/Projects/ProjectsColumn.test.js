import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import ProjectsColumn from './ProjectsColumn';
import Project from './Project';

const mockProjects = [
    {
        url: 'Mock URL',
        name: 'Test Project Name',
        covers: {
            original: 'Original cover URL'
        }
    },
    {
        url: 'Mock URL 2',
        name: 'Test Project Name 2',
        covers: {
            original: 'Original cover URL 2'
        }
    }
];

const mockProjectsOverFive = [{ url: 'Mock Url', name: 'Test Project Name', covers: { original: 'Original Cover URL' }},{ url: 'Mock Url', name: 'Test Project Name', covers: { original: 'Original Cover URL' }},{ url: 'Mock Url', name: 'Test Project Name', covers: { original: 'Original Cover URL' }},{ url: 'Mock Url', name: 'Test Project Name', covers: { original: 'Original Cover URL' }},{ url: 'Mock Url', name: 'Test Project Name', covers: { original: 'Original Cover URL' }},{ url: 'Mock Url', name: 'Test Project Name', covers: { original: 'Original Cover URL' }},{ url: 'Mock Url', name: 'Test Project Name', covers: { original: 'Original Cover URL' }},{ url: 'Mock Url', name: 'Test Project Name', covers: { original: 'Original Cover URL' }}];

describe('ProjectsColumn', () => {
    it('snapshot: should render correctly', () => {
        const output = shallow(<ProjectsColumn projects={mockProjects} />);
        expect(shallowToJson(output)).toMatchSnapshot();
    });
    it('snapshot: should not error out without input', () => {
        const output = shallow(<ProjectsColumn />);
        expect(shallowToJson(output)).toMatchSnapshot();
    });
    it('unit: displays a "View all..." link only if there are more than five users', () => {
        const mounted = mount(<ProjectsColumn projects={mockProjectsOverFive} />);
        expect(mounted.containsMatchingElement(<a>View all...</a>)).toBe(true);
        mounted.unmount();
        const mountedLessThan = mount(<ProjectsColumn projects={mockProjects} />);
        expect(mountedLessThan.containsMatchingElement(<a>View all...</a>)).toBe(false);
        mountedLessThan.unmount();
    });
    it('unit: displays a placeholder if no projects are found', () => {
        const mounted = mount(<ProjectsColumn projects={[]} />);
        expect(mounted.containsMatchingElement(<div className='has-text-centered'> This user does not have any projects. </div>)).toBe(true);
        mounted.unmount();
        const mounted2 = mount(<ProjectsColumn />);
        expect(mounted2.containsMatchingElement(<div className='has-text-centered'> This user does not have any projects. </div>)).toBe(true);
        mounted2.unmount();
        const mounted3 = mount(<ProjectsColumn projects={mockProjects} />);
        expect(mounted3.containsMatchingElement(<div className='has-text-centered'> This user does not have any projects. </div>)).toBe(false);
        mounted3.unmount();
    });
    it('unit: only displays five or less projects at a time', () => {
        const mounted = mount(<ProjectsColumn projects={mockProjectsOverFive} />);
        expect(mounted.find('div.column').length).toBe(5);
        mounted.unmount();
        const mounted2 = mount(<ProjectsColumn projects={mockProjects} />);
        expect(mounted2.find('div.column').length).toBe(mockProjects.length);
        mounted2.unmount();
    })
});
