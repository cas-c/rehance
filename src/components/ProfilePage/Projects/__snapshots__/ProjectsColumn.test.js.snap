// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectsColumn snapshot: should not error out without input 1`] = `
<div>
  <div
    className="columns"
  >
    <div
      className="has-text-centered"
    >
       This user does not have any projects. 
    </div>
  </div>
</div>
`;

exports[`ProjectsColumn snapshot: should render correctly 1`] = `
<div>
  <div
    className="columns"
  >
    <div
      className="column"
      key="project-0-wrapper"
    >
      <Project
        data={
          Object {
            "covers": Object {
              "original": "Original cover URL",
            },
            "name": "Test Project Name",
            "url": "Mock URL",
          }
        }
        key="project-0"
      />
    </div>
    <div
      className="column"
      key="project-1-wrapper"
    >
      <Project
        data={
          Object {
            "covers": Object {
              "original": "Original cover URL 2",
            },
            "name": "Test Project Name 2",
            "url": "Mock URL 2",
          }
        }
        key="project-1"
      />
    </div>
  </div>
</div>
`;
